addons/mail_client_extension/controllers/main.py

```diff
@@ -0,0 +1,261 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+import base64
+import datetime
+import hashlib
+import hmac
+import json
+import logging
+import odoo
+import werkzeug
+
+from odoo import http, tools
+from odoo.http import request
+
+_logger = logging.getLogger(__name__)
+
+
+class MailClientExtensionController(http.Controller):
+
+    @http.route('/mail_client_extension/auth', type='http', auth="user", methods=['GET'], website=True)
+    def auth(self, **values):
+        """
+         Once authenticated this route renders the view that shows an app wants to access Odoo.
+         The user is invited to allow or deny the app. The form posts to `/mail_client_extension/auth/confirm`.
+         """
+        return request.render('mail_client_extension.app_auth', values)
+
+    @http.route('/mail_client_extension/auth/confirm', type='http', auth="user", methods=['POST'])
+    def auth_confirm(self, scope, friendlyname, redirect, info=None, do=None, **kw):
+        """
+        Called by the `app_auth` template. If the user decided to allow the app to access Odoo, a temporary auth code
+        is generated and he is redirected to `redirect` with this code in the URL. It should redirect to the app, and
+        the app should then exchange this auth code for an access token by calling
+        `/mail_client_extension/auth/access_token`.
+        """
+        parsed_redirect = werkzeug.urls.url_parse(redirect)
+        params = parsed_redirect.decode_query()
+        if do:
+            name = friendlyname if not info else f'{friendlyname}: {info}'
+            auth_code = self._generate_auth_code(scope, name)
+            params.update({'success': 1, 'auth_code': auth_code})
+        else:
+            params.update({'success': 0})
```

Comment:
nbd but `update()` can take kwargs, and it's overly complicated to use `update` to update a single key unless otherwise constrained:
```suggestion
            params.update(success: 1, auth_code: auth_code)
        else:
            params['success'] = 0
```